sdk/attestation/azure-security-attestation/inc/azure/attestation/attestation_administration_client.hpp

```diff
@@ -44,9 +44,108 @@ namespace Azure { namespace Security { namespace Attestation {
    *
    */
   class AttestationAdministrationClient final {
-    friend class AttestationAdministrationClientFactory;
+    friend class AttestationAdministrationClientCreator;
 
   public:
+    /** @brief Intermediate factory result that allows factory methods to return both a pointer and
+     * a concrete class implementation.
+     *
+     * See [Raymond Chen's Blob
+     * Post](https://devblogs.microsoft.com/oldnewthing/20191106-00/?p=103066) for more information.
+     */
+    class AttestationAdministrationClientCreator {
```

Comment:
This class should not be public.

Also, when you are declaring a friend this way, you mean the class that is in the same level of hierarchy as client and not a nested class with the same name. The reason the nested class has access to private members is that all nested classes have such access to the parents. Friend declaration has no effect in this case, it is not causing it.

Can we do the following?:
**/inc/azure/attestation/internal/attestation_client_factory.hpp**
```
#inlcude "azure/attestation/attestation_client.hpp"
#include "azure/attestation/attestation_administration_client.hpp"
#include <memory>
namespace Azure { namespace Security { namespace Attestation { namespace _detail {
// it is _detail namespace and not _internal, because this is not meant to be used by other packages
class AttestationFactory { // name is not important
public:
    static smart_ptr<AttestationClient> CreateAttestationClient(...)
    static smart_ptr<AttestationAdministrationClient> CreateAttestationAdministrationClient(...)
};
} } } }
```
**/inc/azure/attestation/attestation_client.hpp AND the same for administration client**
```
namespace Azure { namespace Security { namespace Attestation {
    namespace _detail {
        class AttestationFactory; // forward declaration
    }
class AttestationClient { 
    friend _detail::AttestationFactory;
public:
    static AttestationClient Create(...);
};
} } } }
```

**tests.cpp**
```
#inlcude "attestation/internal/attestation_client_factory.hpp"

...
auto client = Azure::Security::Attestation::_detail::AttestationFactory::CreateAttestationClient(...);
```